src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -266,17 +266,156 @@ tx_gateway_frontend::find_coordinator(kafka::transactional_id id) {
     co_return x;
 }
 
-std::optional<model::ntp>
+ss::future<std::optional<model::ntp>>
 tx_gateway_frontend::get_ntp(kafka::transactional_id id) {
-    auto tx_ntp_opt = _tx_coordinator_ntp_mapper.local().ntp_for(
-      kafka::group_id(id));
+    auto cfg = _metadata_cache.local().get_topic_cfg(model::tx_manager_nt);
+    if (!cfg) {
+        // Transaction coordinator topic not exist in cache
+        // should be catched by caller (find_coordinator)
+        // It must wait for topic in cache or init topic
+        co_return std::nullopt;
+    }
+    int32_t partitions_amount = cfg->partition_count;
+    for (auto p = 0; p < partitions_amount; ++p) {
+        auto hosted_by_partition = co_await hosts(model::partition_id(p), id);
+        if (hosted_by_partition.has_error() || !hosted_by_partition.value()) {
+            continue;
+        } else {
+            co_return model::ntp(
+              model::tx_manager_nt.ns,
+              model::tx_manager_nt.tp,
+              model::partition_id(p));
+        }
+    }
+    auto tx_ntp_opt = co_await _tx_coordinator_ntp_mapper.local().ntp_for(
+      id, _controller->self());
     if (!tx_ntp_opt) {
         vlog(
           txlog.debug,
           "Topic {} doesn't exist in metadata cache",
           model::tx_manager_nt);
     }
-    return tx_ntp_opt;
+    co_return tx_ntp_opt;
+}
+
+ss::future<checked<bool, tx_errc>> tx_gateway_frontend::hosts(
+  model::partition_id partition, kafka::transactional_id tx_id) {
+    model::ntp tx_ntp(
+      model::tx_manager_nt.ns, model::tx_manager_nt.tp, partition);
+    auto retries = _metadata_dissemination_retries;
+    auto delay_ms = _metadata_dissemination_retry_delay_ms;
+    auto aborted = false;
+
+    auto leader_opt = _leaders.local().get_leader(tx_ntp);
+    while (!aborted && !leader_opt && 0 < retries--) {
+        vlog(
+          txlog.trace,
+          "waiting for {} to fill leaders cache, retries left: {}",
+          tx_ntp,
+          retries);
+        aborted = !co_await sleep_abortable(delay_ms, _as);
+        leader_opt = _leaders.local().get_leader(tx_ntp);
+    }
+    if (!leader_opt) {
+        vlog(txlog.warn, "can't find {} in the leaders cache", tx_ntp);
+        co_return tx_errc::leader_not_found;
+    }
+
+    auto leader = leader_opt.value();
+    auto _self = _controller->self();
+
+    if (leader == _self) {
+        co_return co_await do_hosts(partition, tx_id);
+    }
+
+    co_return tx_errc::not_coordinator;
+}
+
+ss::future<checked<bool, tx_errc>> tx_gateway_frontend::do_hosts(
```

Comment:
do_host should:

- take read lock, sync and check if an stm hosts a tx (grabbing write lock first is the perf killer)
- return true if it hosts
- return false on error
- on false:
  - release read lock
  - attempt do_init_hosted_transactions (assuming it take write lock)
  - on error return false
  - check if an stm hosts a tx
  - sync isn't necessary because do_init_hosted_transactions waits for the state to be applied